assume/markets/clearing_algorithms/contracts.py

```diff
@@ -0,0 +1,659 @@
+# SPDX-FileCopyrightText: ASSUME Developers
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from itertools import groupby
+from operator import itemgetter
+from typing import Callable
+
+import pandas as pd
+from dateutil import rrule as rr
+from dateutil.relativedelta import relativedelta as rd
+
+from assume.common.market_objects import (
+    ClearingMessage,
+    MarketConfig,
+    MarketProduct,
+    MetaDict,
+    Order,
+    Orderbook,
+)
+from assume.markets.base_market import MarketRole
+
+log = logging.getLogger(__name__)
+
+
+class PayAsBidContractRole(MarketRole):
+    required_fields = [
+        "sender_id",
+        "contract",
+        "eligible_lambda",
+        "evaluation_frequency",
+    ]
+
+    def __init__(
+        self,
+        marketconfig: MarketConfig,
+        limitation: str = "only_co2emissionless",
+        market_indices: dict[str, str] = {},
+    ):
+        super().__init__(marketconfig)
+        self.limitation = limitation
+        self.market_indices = market_indices
+        self.futures = {}
+
+    def setup(self):
+        super().setup()
+
+        def accept_data_response(content: dict, meta: MetaDict):
+            return content.get("context") == "data_response"
+
+        self.context.subscribe_message(
+            self, self.handle_data_response, accept_data_response
+        )
+
+    def handle_data_response(self, content: dict, meta: MetaDict) -> None:
+        """
+        Handles the response data and finishes the async Future waiting for the data.
+
+        Args:
+            content (dict): message content with data property
+            meta (MetaDict): message meta
+        """
+        if meta["in_reply_to"] not in self.futures:
+            log.error(f'data response {meta["in_reply_to"]} not in awaited futures')
+        else:
+            self.futures[meta["in_reply_to"]].set_result(content["data"])
+
+    def validate_registration(self, content: dict, meta: MetaDict) -> bool:
+        """
+        validation function called by handle_registration
+        Makes it possible to allow only a subset of agents to bid on this market
+        by using self.limitation of the clearing mechanism.
+
+        Args:
+            content (dict): message content with registration message and agent information
+            meta (MetaDict): message meta
+
+        Returns:
+            bool: True if agent fulfills requirements
+        """
+        if self.limitation:
+            if self.limitation == "only_co2emissionless":
```

Comment:
maybe we could make these technologies as a part of the config instead of hard coding them at this spot?

Reply:
I wanted to improve this once we have more than 2 possible limitations of market participants.
What do you mean with "part of the config"?

Reply:
since now we have additional configs for each market, we could list such technologies there

Reply:
Good point.
But I don't know if this will only stay "a list of allowed technologies" - one could also specifiy to limit this market to agents with a max_power > 10MW or whatever - so I would like to keep this loose for now